packages/jazz-tools/src/implementation/zodSchema/unionUtils.ts

```diff
@@ -46,49 +49,52 @@ export function schemaUnionDiscriminatorFor(
           "z.discriminatedUnion() of collaborative types is not supported for CoLists",
         );
       }
-      let remainingOptions = [...flattendOptions];
-      for (const discriminator of schema._zod.disc.keys()) {
-        const discriminatorValue = (_raw as RawCoMap).get(
-          discriminator as string,
-        );
-        if (typeof discriminatorValue !== "string") {
-          throw new Error("Discriminator must be a string");
+
+      for (const option of flattendOptions) {
+        if (option._zod.def.type !== "object") {
+          continue;
         }
-        remainingOptions = remainingOptions.filter((option) => {
-          if (option._zod.def.type !== "object") {
-            return false;
-          }
+
+        let match = true;
+        for (const key of schema._zod.disc.keys()) {
           const discriminatorDef = (option as z.core.$ZodObject)._zod.def.shape[
-            discriminator as string
+            key as string
           ];
+
+          const discriminatorValue = (_raw as RawCoMap).get(key as string);
+
+          if (discriminatorValue && typeof discriminatorValue === "object") {
+            throw new Error("Discriminator must be a primitive value");
+          }
+
           if (!discriminatorDef) {
-            return false;
+            if (key === discriminator) {
+              match = false;
+              break;
+            } else {
+              continue;
+            }
           }
 
           if (discriminatorDef._zod.def.type !== "literal") {
-            console.warn(
-              "Non-literal discriminator found in z.discriminatedUnion() of collaborative types",
-            );
-            return false;
+            break;
           }
-          if (
-            (discriminatorDef._zod.def as z.core.$ZodLiteralDef).values
-              .length !== 1
-          ) {
-            console.warn(
-              "Literal discriminator with more than one value found in z.discriminatedUnion() of collaborative types",
-            );
-            return false;
+
+          const literalDef = discriminatorDef._zod.def as z.core.$ZodLiteralDef;
+
+          for (const value of literalDef.values) {
+            if (value !== discriminatorValue) {
+              match = false;
+              break;
+            }
           }
-          return (
-            (discriminatorDef._zod.def as z.core.$ZodLiteralDef).values[0] ===
-            discriminatorValue
-          );
-        });
-        if (remainingOptions.length === 1) {
-          return zodSchemaToCoSchema(remainingOptions[0]!);
+        }
+
+        if (match) {
```

Comment:
The first field that matches all the discriminators passes.

Any non-defined implicit discriminator is skipped in the check